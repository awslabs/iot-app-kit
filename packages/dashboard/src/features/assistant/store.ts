import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import uniqBy from 'lodash-es/uniqBy';
import { v4 as uuid } from 'uuid';
import type { AssistantStateTypes } from '#types';

export interface AssistantState {
  state: AssistantStateTypes;
  conversationId: string;
  isChatbotOpen: boolean;
  callerComponentId?: string;
  action?: string;
  actionId?: string;
  mode: 'on' | 'off';
  selectedQueries: {
    widgetId: string;
    widgetType: string;
    selectedProperties: number;
  }[];
}

export const initialState: AssistantState = {
  state: 'PASSIVE',
  conversationId: uuid(),
  isChatbotOpen: false,
  mode: 'off',
  selectedQueries: [],
};

export const {
  reducer,
  actions: {
    toggleAssistantMode,
    toggleChatbot,
    cleanAssistant,
    assistantSelectWidgets,
    assistantDeselectWidgets,
    assistantCleanWidgetsSelection,
  },
} = createSlice({
  name: 'assistant',
  initialState: initialState,
  reducers: {
    toggleChatbot: (
      state,
      {
        payload,
      }: PayloadAction<{
        open: boolean;
        callerComponentId: string;
        action?: 'summarize' | 'divedeep';
      }>
    ) => {
      state.isChatbotOpen = payload.open;
      state.callerComponentId = payload.callerComponentId;
      state.action = payload.action;
      state.actionId = uuid();
    },

    toggleAssistantMode: (
      state,
      { payload }: PayloadAction<{ mode: 'on' | 'off' }>
    ) => {
      state.mode = payload.mode;
    },

    cleanAssistant: (state) => {
      state.conversationId = uuid();
      state.callerComponentId = undefined;
      state.action = undefined;
      state.selectedQueries = [];
    },

    assistantSelectWidgets: (
      state,
      {
        payload,
      }: PayloadAction<{
        widgetId: string;
        widgetType: string;
        selectedProperties: number;
      }>
    ) => {
      state.selectedQueries = uniqBy(
        [...state.selectedQueries, payload],
        'widgetId'
      );
    },

    assistantDeselectWidgets: (
      state,
      { payload }: PayloadAction<{ widgetId: string }>
    ) => {
      state.selectedQueries = state.selectedQueries.filter(
        ({ widgetId }) => widgetId !== payload.widgetId
      );
    },

    assistantCleanWidgetsSelection: (state) => {
      state.selectedQueries = [];
    },
  },
});
