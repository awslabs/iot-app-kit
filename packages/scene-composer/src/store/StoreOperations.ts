export type OperationType = 'INITIALIZE' | 'UPDATE_DOCUMENT' | 'UPDATE_EDITOR' | 'TRANSIENT';

export type SceneComposerEditorOperation =
  | 'resetEditorState'
  | 'setCameraControlsType'
  | 'setEditorConfig'
  | 'setSelectedSceneNodeRef'
  | 'setSubModelSelection'
  | 'setHighlightedSceneNodeRef'
  | 'setLoadingModelState'
  | 'setTransformControlMode'
  | 'setTransformControls'
  | 'setCameraTarget'
  | 'addMessages'
  | 'clearMessages'
  | 'setAddingWidget'
  | 'setCursorPosition'
  | 'setCursorLookAt'
  | 'setCursorVisible'
  | 'setCursorStyle'
  | 'setActiveCameraSettings'
  | 'setActiveCameraName'
  | 'setMainCameraObject';

export type SceneComposerDocumentOperation =
  | 'loadScene'
  | 'appendSceneNodeInternal'
  | 'updateSceneNodeInternal'
  | 'updateSceneNodeInternalTransient'
  | 'updateDocumentInternal'
  | 'removeSceneNode'
  | 'updateSceneRuleMapById'
  | 'removeSceneRuleMapById'
  | 'addComponentInternal'
  | 'updateComponentInternal'
  | 'removeComponent'
  | 'setSceneProperty'
  | 'clearTemplatizedDataBindings';

export type SceneComposerDataOperation = 'setDataInput' | 'setDataBindingTemplate';

export type SceneComposerViewOptionOperation = 'toggleMotionIndicatorVisibility' | 'setTagSettings';

export type SceneComposerOperation =
  | SceneComposerEditorOperation
  | SceneComposerDocumentOperation
  | SceneComposerViewOptionOperation
  | SceneComposerDataOperation;

export const SceneComposerOperationTypeMap: Record<SceneComposerOperation, OperationType> = {
  // loadScene is a speical operation as we want to clear the undo/redo state after this
  loadScene: 'INITIALIZE',

  appendSceneNodeInternal: 'UPDATE_DOCUMENT',
  updateSceneNodeInternal: 'UPDATE_DOCUMENT',
  updateSceneNodeInternalTransient: 'TRANSIENT',
  updateDocumentInternal: 'UPDATE_DOCUMENT',
  removeSceneNode: 'UPDATE_DOCUMENT',
  addComponentInternal: 'UPDATE_DOCUMENT',
  updateComponentInternal: 'UPDATE_DOCUMENT',
  removeComponent: 'UPDATE_DOCUMENT',
  updateSceneRuleMapById: 'UPDATE_DOCUMENT',
  removeSceneRuleMapById: 'UPDATE_DOCUMENT',
  setSceneProperty: 'UPDATE_DOCUMENT',
  clearTemplatizedDataBindings: 'UPDATE_DOCUMENT',

  resetEditorState: 'UPDATE_EDITOR',
  setCameraControlsType: 'UPDATE_EDITOR',
  setEditorConfig: 'UPDATE_EDITOR',
  setSelectedSceneNodeRef: 'UPDATE_EDITOR',
  setSubModelSelection: 'UPDATE_EDITOR',
  setHighlightedSceneNodeRef: 'UPDATE_EDITOR',
  setTransformControlMode: 'UPDATE_EDITOR',
  setTransformControls: 'UPDATE_EDITOR',
  setCameraTarget: 'UPDATE_EDITOR',
  addMessages: 'UPDATE_EDITOR',
  clearMessages: 'UPDATE_EDITOR',
  setAddingWidget: 'UPDATE_EDITOR',
  setCursorPosition: 'UPDATE_EDITOR',
  setCursorLookAt: 'UPDATE_EDITOR',
  setCursorVisible: 'UPDATE_EDITOR',
  setCursorStyle: 'UPDATE_EDITOR',
  setActiveCameraSettings: 'UPDATE_EDITOR',
  setActiveCameraName: 'UPDATE_EDITOR',
  setMainCameraObject: 'UPDATE_EDITOR',

  setLoadingModelState: 'TRANSIENT',
  setDataInput: 'TRANSIENT',
  setDataBindingTemplate: 'TRANSIENT',

  toggleMotionIndicatorVisibility: 'TRANSIENT',
  setTagSettings: 'TRANSIENT',
};
