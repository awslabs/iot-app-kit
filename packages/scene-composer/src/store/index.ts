export * from './StoreOperations';
export * from './Store';
