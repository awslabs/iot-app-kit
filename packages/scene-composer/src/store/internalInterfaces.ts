import { IErrorDetails } from '../common/errors';
import {
  SelectionChangedEventCallback,
  IAnchorComponent,
  ICameraComponent,
  IColorOverlayComponent,
  ILightComponent,
  IModelRefComponent,
  IMotionIndicatorComponent,
  IRuleBasedMap,
  IRuleStatement,
  ISceneComponent,
  ISceneDocument,
  ISceneNode,
  ITransform,
  ITransformConstraint,
  IValueDataBinding,
  IValueDataBindingProvider,
  OperationMode,
  ShowAssetBrowserCallback,
  URIModifier,
  KnownComponentType,
  WidgetClickEventCallback,
  ISubModelRefComponent,
  IDataOverlayComponent,
} from '../interfaces';
import { MapControls as MapControlsImpl, OrbitControls as OrbitControlsImpl } from '../three/OrbitControls';

export type ISerializationErrorDetails = IErrorDetails;

export interface IDeserializationResult {
  document?: ISceneDocumentInternal;
  errors: ISerializationErrorDetails[];
}

/******************************************************************************
 * Editor Data Model
 ******************************************************************************/

export enum DisplayMessageCategory {
  Error = 'Error',
  Warning = 'Warning',
  Info = 'Info',
}

export type CameraControlImpl = MapControlsImpl | OrbitControlsImpl;
export type TweenValueObject = { x: number; y: number; z: number };

export type CursorStyle = 'move' | 'edit';

export interface IDisplayMessage {
  category: DisplayMessageCategory;
  messageText: string;
  params?: Record<string, string>;
}

export interface IEditorConfig {
  operationMode?: OperationMode;
  uriModifier?: URIModifier;
  valueDataBindingProvider?: IValueDataBindingProvider;
  showAssetBrowserCallback?: ShowAssetBrowserCallback;
  onWidgetClick?: WidgetClickEventCallback;
  onSelectionChanged?: SelectionChangedEventCallback;
}

/******************************************************************************
 * Document Model
 ******************************************************************************/

export type ITransformInternal = ITransform;
export type IRuleBasedMapInternal = IRuleBasedMap;
export type IRuleStatementInternal = IRuleStatement;

export interface ISceneNodeInternal extends ISceneNode {
  // below fields are not nullable internally
  ref: string;
  name: string;
  transform: ITransformInternal;
  transformConstraint: ITransformConstraint;
  components: ISceneComponentInternal[];
  childRefs: string[];
  properties: Record<string, any>;
}

export interface ISceneDocumentInternal extends ISceneDocument {
  unit: string; // unit is not nullable internally, default to 'meter'
  ruleMap: Record<string, IRuleBasedMapInternal>;
  nodeMap: Record<string, ISceneNodeInternal>;
  componentNodeMap: { [type in KnownComponentType | string]?: Record<string, string[]> }; // record of node.ref to component.ref[]
}

export interface ISceneComponentInternal extends ISceneComponent {
  // unique id for the component for internal reference
  ref: string;
}

export interface IDataBoundSceneComponentInternal extends ISceneComponentInternal {
  valueDataBinding?: IValueDataBinding;
}

/******************************************************************************
 * Components
 ******************************************************************************/

export type ILightComponentInternal = ISceneComponentInternal & ILightComponent;

export type ICameraComponentInternal = ISceneComponentInternal & ICameraComponent;

export type IModelRefComponentInternal = ISceneComponentInternal & IModelRefComponent;

export type ISubModelRefComponentInternal = ISceneComponentInternal & ISubModelRefComponent;

export type IAnchorComponentInternal = IDataBoundSceneComponentInternal & IAnchorComponent;

export type IColorOverlayComponentInternal = IDataBoundSceneComponentInternal & IColorOverlayComponent;

export type IMotionIndicatorComponentInternal = ISceneComponentInternal & IMotionIndicatorComponent;

export type IDataOverlayComponentInternal = ISceneComponentInternal & IDataOverlayComponent;

/******************************************************************************
 * Type magic...
 ******************************************************************************/

export const isISceneNodeInternal = (item: ISceneNodeInternal | undefined): item is ISceneNodeInternal => {
  return !!item;
};

export const isISceneComponentInternal = (
  item: ISceneComponentInternal | undefined,
): item is ISceneComponentInternal => {
  return !!item;
};
