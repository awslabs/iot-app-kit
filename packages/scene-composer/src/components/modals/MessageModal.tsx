import _ from 'lodash';
import { Box, Button, Header, Icon, SpaceBetween, StatusIndicator, TextContent } from '@awsui/components-react';
import React, { Fragment, useContext, useMemo } from 'react';
import { useIntl } from 'react-intl';

import { AWS_USER_GUIDE_DOC_URL } from '../../common/constants';
import { sceneComposerIdContext } from '../../common/sceneComposerIdContext';
import { useStore } from '../../store';
import { DisplayMessageCategory } from '../../store/internalInterfaces';
import { evalStringTemplate } from '../../utils/stringUtils';
import CenteredContainer from '../CenteredContainer';

function mapMessageCategoryToModalTitle(category: DisplayMessageCategory) {
  const intl = useIntl();
  switch (category) {
    case DisplayMessageCategory.Error:
      return intl.formatMessage({ defaultMessage: 'Error - unable to render the scene', description: 'Error message' });
    case DisplayMessageCategory.Warning:
      return intl.formatMessage({ defaultMessage: 'Warning', description: 'Warning message' });
    default:
      return intl.formatMessage({ defaultMessage: 'Notice', description: 'Notice message' });
  }
}

const MessageModal = () => {
  const sceneComposerId = useContext(sceneComposerIdContext);
  const messages = useStore(sceneComposerId)((state) => state.getMessages());
  const clearMessages = useStore(sceneComposerId)((state) => state.clearMessages);
  const intl = useIntl();

  const containsError = messages.find((m) => m.category === DisplayMessageCategory.Error);
  const containsWarning = messages.find((m) => m.category === DisplayMessageCategory.Warning);

  const title = containsError
    ? mapMessageCategoryToModalTitle(DisplayMessageCategory.Error)
    : containsWarning
    ? mapMessageCategoryToModalTitle(DisplayMessageCategory.Warning)
    : mapMessageCategoryToModalTitle(DisplayMessageCategory.Info);
  const allowDismiss = !containsError;
  const showUserGuideLink = containsError || containsWarning;

  const messageContent = useMemo(() => {
    const groupedMessages = _.groupBy(messages, (m) => m.category);
    return [DisplayMessageCategory.Error, DisplayMessageCategory.Warning, DisplayMessageCategory.Info]
      .map((category) => {
        if (category in groupedMessages) {
          const messages = groupedMessages[category];
          if (messages.length === 0) {
            return null;
          }
          return (
            <Fragment key={category}>
              {messages.map((m, index) => (
                <StatusIndicator key={index} type={category.toLocaleLowerCase() as any}>
                  {evalStringTemplate(m.messageText, m.params)}
                  {m.params?.cause ? ' Cause - ' + m.params.cause : ''}
                </StatusIndicator>
              ))}
            </Fragment>
          );
        } else {
          return null;
        }
      })
      .filter((m) => m !== null);
  }, [messages]);

  return (
    <CenteredContainer
      header={<Header variant='h2'>{title}</Header>}
      footer={
        allowDismiss && (
          <Box float='right' padding={{ bottom: 's' }}>
            <Button variant='primary' onClick={clearMessages}>
              Ok
            </Button>
          </Box>
        )
      }
    >
      <SpaceBetween size='s'>
        <Box>{messageContent}</Box>

        {showUserGuideLink && (
          <Box>
            <TextContent>
              <p>
                <a href={AWS_USER_GUIDE_DOC_URL}>
                  {' '}
                  {intl.formatMessage({ defaultMessage: 'Learn More', description: 'Hyperlink title' })}{' '}
                  <Icon name='external' size='normal' variant='link' />{' '}
                </a>
              </p>
            </TextContent>
          </Box>
        )}
      </SpaceBetween>
    </CenteredContainer>
  );
};

MessageModal.displayName = 'MessageModal';

export default MessageModal;
