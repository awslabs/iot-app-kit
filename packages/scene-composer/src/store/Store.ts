import create, { StateCreator, UseStore } from 'zustand';
import shallow from 'zustand/shallow';

import { log, immer, undoMiddleware, UndoState } from './middlewares';
import { SceneComposerOperation } from './StoreOperations';
import { createSceneDocumentSlice, ISceneDocumentSlice } from './slices/SceneDocumentSlice';
import { createEditStateSlice, IEditorStateSlice } from './slices/EditorStateSlice';
import { IDataStoreSlice, createDataStoreSlice } from './slices/DataStoreSlice';
import { createNodeErrorStateSlice, INodeErrorStateSlice } from './slices/NodeErrorStateSlice';
import { createViewOptionStateSlice, IViewOptionStateSlice } from './slices/ViewOptionStateSlice';
import {
  ISceneDocumentInternal,
  ISceneNodeInternal,
  ISceneComponentInternal,
  IModelRefComponentInternal,
  ISubModelRefComponentInternal,
  ICameraComponentInternal,
  IAnchorComponentInternal,
  ILightComponentInternal,
  IColorOverlayComponentInternal,
  isISceneComponentInternal,
  isISceneNodeInternal,
  IMotionIndicatorComponentInternal,
} from './internalInterfaces';

export type {
  // Document
  ISceneDocumentInternal,
  ISceneNodeInternal,
  ISceneComponentInternal,
  // Components
  IModelRefComponentInternal,
  ISubModelRefComponentInternal,
  ICameraComponentInternal,
  IAnchorComponentInternal,
  ILightComponentInternal,
  IColorOverlayComponentInternal,
  IMotionIndicatorComponentInternal,
};

export interface ISharedState {
  lastOperation?: SceneComposerOperation;
  noHistoryStates: IViewOptionStateSlice;
}

export type RootState = ISharedState &
  ISceneDocumentSlice &
  IEditorStateSlice &
  IDataStoreSlice &
  UndoState &
  INodeErrorStateSlice;

/**
 * Core state management functions
 * TODO: make them into slices and better organized
 */
const stateCreator: StateCreator<RootState> = (set, get, api) => ({
  lastOperation: undefined,
  ...createSceneDocumentSlice(set, get, api),
  ...createEditStateSlice(set, get, api),
  ...createDataStoreSlice(set, get, api),
  noHistoryStates: {
    ...createViewOptionStateSlice(set, get, api),
  },
  ...createNodeErrorStateSlice(set, get, api),
});

const createStateImpl: () => UseStore<RootState> = () => create<RootState>(undoMiddleware(log(immer(stateCreator))));

// TODO: currently undoMiddleware will record editor state changes, such as select/deselect object.
// We may want to fine-tune the undo/redo experience.
const stores = new Map<string, UseStore<RootState>>();

const useStore: (id: string) => UseStore<RootState> = (id: string) => {
  if (!stores.has(id)) {
    stores.set(id, createStateImpl());
  }
  return stores.get(id)!;
};

const sceneDocumentSelector = (state: RootState) => ({
  document: state.document,
  sceneLoaded: state.sceneLoaded,
  getSceneNodeByRef: state.getSceneNodeByRef,
  getSceneNodesByRefs: state.getSceneNodesByRefs,
  appendSceneNodeInternal: state.appendSceneNodeInternal,
  updateSceneNodeInternal: state.updateSceneNodeInternal,
  updateDocumentInternal: state.updateDocumentInternal,
  listSceneRuleMapIds: state.listSceneRuleMapIds,
  getSceneRuleMapById: state.getSceneRuleMapById,
  updateSceneRuleMapById: state.updateSceneRuleMapById,
  removeSceneRuleMapById: state.removeSceneRuleMapById,
  getSceneProperty: state.getSceneProperty,
  removeSceneNode: state.removeSceneNode,
});

const editorStateSelector = (state: RootState) => ({
  editorConfig: state.editorConfig,
  isViewing: state.isViewing,
  isEditing: state.isEditing,
  addingWidget: state.addingWidget,
  isLoadingModel: state.isLoadingModel,
  transformControls: state.transformControls,
  transformControlMode: state.transformControlMode,
  cameraCommand: state.cameraCommand,
  cameraControlsType: state.cameraControlsType,
  selectedSceneNodeRef: state.selectedSceneNodeRef,
  selectedSceneSubmodelRef: state.selectedSceneSubmodelRef,
  cursorPosition: state.cursorPosition,
  cursorLookAt: state.cursorLookAt,
  cursorVisible: state.cursorVisible,
  cursorStyle: state.cursorStyle,

  setEditorConfig: state.setEditorConfig,
  getObject3DBySceneNodeRef: state.getObject3DBySceneNodeRef,
  setSelectedSceneNodeRef: state.setSelectedSceneNodeRef,
  setSelectedSceneSubmodelRef: state.setSelectedSceneSubmodelRef,
  setTransformControls: state.setTransformControls,
  setTransformControlMode: state.setTransformControlMode,
  setCameraTarget: state.setCameraTarget,
  setCameraControlsType: state.setCameraControlsType,
  setSceneNodeObject3DMapping: state.setSceneNodeObject3DMapping,
  setAddingWidget: state.setAddingWidget,
  setCursorPosition: state.setCursorPosition,
  setCursorLookAt: state.setCursorLookAt,
  setCursorVisible: state.setCursorVisible,
  setCursorStyle: state.setCursorStyle,
  activeCameraSettings: state.activeCameraSettings,
  setActiveCameraSettings: state.setActiveCameraSettings,
  activeCameraName: state.activeCameraName,
  setActiveCameraName: state.setActiveCameraName,
  mainCameraObject: state.mainCameraObject,
  setMainCameraObject: state.setMainCameraObject,
});

const dataStoreSelector = (state: RootState) => ({
  dataBindingTemplate: state.dataBindingTemplate,
  dataInput: state.dataInput,
  setDataInput: state.setDataInput,
});

const nodeErrorStateSelector = (state: RootState) => ({
  nodeErrorMap: state.nodeErrorMap,
  addNodeError: state.addNodeError,
  removeNodeError: state.removeNodeError,
});

const viewOptionStateSelector = (state: RootState) => ({
  motionIndicatorVisible: state.noHistoryStates.motionIndicatorVisible,
  toggleMotionIndicatorVisibility: state.noHistoryStates.toggleMotionIndicatorVisibility,
  tagSettings: state.noHistoryStates.tagSettings,
  setTagSettings: state.noHistoryStates.setTagSettings,
});

/**
 * useSceneDocument is a useful short-hand hook for reacting to scene document changes.
 * NOTE: this will cause refresh whenever there is a change in the whole document.
 * You should use a smaller granular state if that's not your intention.
 */
const useSceneDocument = (id: string) => {
  return useStore(id)(sceneDocumentSelector, shallow);
};

const useEditorState = (id: string) => {
  return useStore(id)(editorStateSelector, shallow);
};

const useDataStore = (id: string) => {
  return useStore(id)(dataStoreSelector, shallow);
};

const useNodeErrorState = (id: string) => {
  return useStore(id)(nodeErrorStateSelector, shallow);
};

const useViewOptionState = (id: string) => {
  return useStore(id)(viewOptionStateSelector, shallow);
};

const isDocumentStateChanged = (current: ISceneDocumentInternal, previous: ISceneDocumentInternal): boolean => {
  // TODO: we'll just implement a simple comparision version for beta release
  return !shallow(current, previous);
};

export {
  useStore,
  sceneDocumentSelector,
  editorStateSelector,
  useSceneDocument,
  useEditorState,
  dataStoreSelector,
  useDataStore,
  nodeErrorStateSelector,
  useNodeErrorState,
  isDocumentStateChanged,
  isISceneComponentInternal,
  isISceneNodeInternal,
  viewOptionStateSelector,
  useViewOptionState,
};
