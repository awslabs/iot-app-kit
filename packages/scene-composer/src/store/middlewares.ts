import { State, StateCreator } from 'zustand';
import { produce, Draft } from 'immer';
import createVanilla, { GetState, SetState, StoreApi } from 'zustand/vanilla';

import DebugLogger from '../logger/DebugLogger';

import { SceneComposerOperationTypeMap } from './StoreOperations';
import { RootState } from './Store';

const LOG = new DebugLogger('stateStore');

/**
 * Log the update to the state store, the prev state and the updated state.
 */
export const log =
  <T extends State>(config: StateCreator<T>): StateCreator<T> =>
  (set, get, api) => {
    return config(
      (...args) => {
        LOG.verbose('old state', get());
        set(...args);
        LOG.verbose('new state', get());
      },
      get,
      api,
    );
  };

/**
 * Make nested state update simple.
 */
export const immer =
  <T extends State>(config: StateCreator<T>): StateCreator<T> =>
  (set, get, api) =>
    config(
      (partial, replace) => {
        const nextState = typeof partial === 'function' ? produce(partial as (state: Draft<T>) => T) : (partial as T);
        set(nextState, replace);
      },
      get,
      api,
    );

/**
 * Undo/Redo, inspired by zundo
 */
export type UndoState = Partial<
  Pick<UndoStoreState, 'undo' | 'redo' | 'clear'> & {
    getUndoState: () => UndoStoreState;
    undoStore: StoreApi<UndoStoreState>;
    // last operation that changed the parent state
    lastOperation?: string;
  }
>;
export interface UndoStoreState {
  prevStates: any[];
  futureStates: any[];
  undo: () => void;
  redo: () => void;
  clear: () => void;
  // handle on the parent store's setter
  setStore: Function;
  // handle on the parent store's getter
  getStore: Function;
}

function filterNoHistoryStates(newState: RootState, currentState: RootState) {
  return {
    ...newState,
    noHistoryStates: currentState.noHistoryStates,
  };
}

// factory to create undoStore. contains memory about past and future states and has methods to traverse states
export const createUndoStore = () => {
  return createVanilla<UndoStoreState>((set, get) => {
    return {
      prevStates: [],
      futureStates: [],
      undo: () => {
        const { prevStates, futureStates, setStore, getStore } = get();
        if (prevStates.length > 0) {
          futureStates.push(getStore());
          const prevState = filterNoHistoryStates(prevStates.pop(), getStore());
          setStore(prevState);
          // force updating the undo store to trigger subscription on the store
          set({ prevStates, futureStates });
        }
      },
      redo: () => {
        const { prevStates, futureStates, setStore, getStore } = get();
        if (futureStates.length > 0) {
          prevStates.push(getStore());
          const futureState = filterNoHistoryStates(futureStates.pop(), getStore());
          setStore(futureState);
          // force updating the undo store to trigger subscription on the store
          set({ prevStates, futureStates });
        }
      },
      clear: () => {
        set({ prevStates: [], futureStates: [] });
      },
      setStore: () => {},
      getStore: () => {},
    };
  });
};

// custom zustand middleware to get previous state
export const undoMiddleware =
  <TState extends UndoState>(config: StateCreator<TState>) =>
  (set: SetState<TState>, get: GetState<TState>, api: StoreApi<TState>) => {
    const undoStore = createUndoStore();
    const { getState, setState } = undoStore;
    const { undo, clear, redo } = getState();
    return config(
      (args) => {
        // Take a snapshot of the current state
        const snapshot = {
          prevStates: [...getState().prevStates, { ...get() }],
          setStore: set,
          futureStates: [],
          getStore: get,
        };

        if (!get().getUndoState) {
          // inject helper functions to user defined store.
          set({
            undo,
            clear,
            redo,
            getUndoState: getState,
            undoStore,
          });
        }

        set(args);

        const lastOperation = get().lastOperation || '';
        const lastOperationType = SceneComposerOperationTypeMap[lastOperation];

        LOG.verbose('lastOperation', lastOperation, 'maps to', lastOperationType);

        if (lastOperationType === 'INITIALIZE') {
          // clear history
          LOG.verbose('clear all history after initialization');
          clear();
        } else if (lastOperationType === 'UPDATE_DOCUMENT') {
          // insert snapshot into the history
          LOG.verbose('insert a new state in undo history', snapshot);
          setState(snapshot);
        } else {
          LOG.verbose("skip the current state as it's not updating the document");
        }
      },
      get,
      api,
    );
  };
