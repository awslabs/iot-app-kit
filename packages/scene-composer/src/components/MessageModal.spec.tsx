import React from 'react';
import { render } from '@testing-library/react';

import { useStore } from '../store';
import { DisplayMessageCategory } from '../store/internalInterfaces';

import MessageModal from './MessageModal';

import Mock = jest.Mock;

jest.mock('../../src/store', () => {
  const originalModule = jest.requireActual('../../src/store');
  return {
    ...originalModule,
    useStore: jest.fn(),
  };
});

describe('MessageModalSnap', () => {
  const clearMessages = jest.fn();
  const mockUseStore = useStore as Mock;

  [
    { category: DisplayMessageCategory.Info, messageText: 'Info Test' },
    { category: DisplayMessageCategory.Warning, messageText: 'Warning Test' },
    { category: DisplayMessageCategory.Error, messageText: 'Error Test' },
  ].map((props) => {
    it(`should render correctly with: ${props}`, () => {
      mockUseStore.mockReturnValueOnce(() => [props]);
      mockUseStore.mockReturnValueOnce(() => clearMessages);

      const { container } = render(<MessageModal />);
      expect(container).toMatchSnapshot();
    });
  });
});
