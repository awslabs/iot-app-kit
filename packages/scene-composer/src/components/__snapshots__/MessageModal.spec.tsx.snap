// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageModalSnap should render correctly with: [object Object] 1`] = `
.c0 {
  width: 600px;
  margin: auto;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  z-index: 1000;
}

<div>
  <div
    class="c0"
    data-mocked="Container"
  >
    <div>
      <div
        data-mocked="Header"
        variant="h2"
      >
        Notice
      </div>
    </div>
    <div>
      <div
        data-mocked="Box"
        float="right"
        padding="{\\"bottom\\":\\"s\\"}"
      >
        <div
          data-mocked="Button"
          variant="primary"
        >
          Ok
        </div>
      </div>
    </div>
    <div
      data-mocked="SpaceBetween"
    >
      <div
        data-mocked="Box"
      >
        <div
          data-mocked="StatusIndicator"
          type="info"
        >
          Info Test
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MessageModalSnap should render correctly with: [object Object] 2`] = `
.c0 {
  width: 600px;
  margin: auto;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  z-index: 1000;
}

<div>
  <div
    class="c0"
    data-mocked="Container"
  >
    <div>
      <div
        data-mocked="Header"
        variant="h2"
      >
        Warning
      </div>
    </div>
    <div>
      <div
        data-mocked="Box"
        float="right"
        padding="{\\"bottom\\":\\"s\\"}"
      >
        <div
          data-mocked="Button"
          variant="primary"
        >
          Ok
        </div>
      </div>
    </div>
    <div
      data-mocked="SpaceBetween"
    >
      <div
        data-mocked="Box"
      >
        <div
          data-mocked="StatusIndicator"
          type="warning"
        >
          Warning Test
        </div>
      </div>
      <div
        data-mocked="Box"
      >
        <div
          data-mocked="TextContent"
        >
          <p>
            <a
              href="https://docs.aws.amazon.com/iot-twinmaker/latest/guide/what-is-twinmaker.html"
            >
               
              Learn More
               
              <div
                data-mocked="Icon"
                name="external"
                variant="link"
              />
               
            </a>
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MessageModalSnap should render correctly with: [object Object] 3`] = `
.c0 {
  width: 600px;
  margin: auto;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  z-index: 1000;
}

<div>
  <div
    class="c0"
    data-mocked="Container"
  >
    <div>
      <div
        data-mocked="Header"
        variant="h2"
      >
        Error - unable to render the scene
      </div>
    </div>
    <div
      data-mocked="SpaceBetween"
    >
      <div
        data-mocked="Box"
      >
        <div
          data-mocked="StatusIndicator"
          type="error"
        >
          Error Test
        </div>
      </div>
      <div
        data-mocked="Box"
      >
        <div
          data-mocked="TextContent"
        >
          <p>
            <a
              href="https://docs.aws.amazon.com/iot-twinmaker/latest/guide/what-is-twinmaker.html"
            >
               
              Learn More
               
              <div
                data-mocked="Icon"
                name="external"
                variant="link"
              />
               
            </a>
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
